File: SentinAL_Crawler/java_crawler/Sentinel.java
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.json.JSONObject;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.MessageDigest; // Added for Hashing
import java.nio.charset.StandardCharsets;

public class Sentinel {
    public static void main(String[] args) {
        // 1. SETUP
        System.setProperty("webdriver.chrome.driver", "chromedriver.exe"); // Ensure this path is correct!

        ChromeOptions options = new ChromeOptions();
        options.addArguments("--headless"); 
        options.addArguments("--disable-gpu");
        // Fix for some environments:
        options.addArguments("--remote-allow-origins=*"); 
        
        WebDriver driver = new ChromeDriver(options);

        try {
            // 2. INGEST
            // NOTE: We will serve this using python -m http.server 8080
            String targetUrl = "http://localhost:8080/pirate-site.html"; 
            System.out.println("[*] Sentinel Crawler Activated. Target: " + targetUrl);

            driver.get(targetUrl);

            String pageSource = driver.getPageSource();
            Document doc = Jsoup.parse(pageSource);

            // 3. EXTRACT (Mapping to Backend Terminology)
            String title = doc.title();
            String uploader = doc.select("meta[name=uploader]").attr("content");
            String location = doc.select("meta[name=server_location]").attr("content");
            String timestamp = doc.select("meta[name=upload_date]").attr("content");
            String durationStr = doc.select("meta[name=duration]").attr("content"); 
            int duration = 0;
            try {
                duration = Integer.parseInt(durationStr);
            } catch (NumberFormatException nfe) {
                duration = 0; // Fallback
        }
            
            // Generate SHA-256 Hash of the HTML source (The "Fingerprint")
            String htmlHash = calculateSHA256(pageSource);

            System.out.println("[+] TARGET ACQUIRED: " + title);
            System.out.println("[+] LOCATION: " + location);
            System.out.println("[+] DURATION: " + duration + " mins");
            System.out.println("[+] EVIDENCE HASH: " + htmlHash.substring(0, 10) + "...");

            // 4. REPORT (Constructing Nested JSON)
            JSONObject innerData = new JSONObject();
            innerData.put("page_title", title); // Matches Backend 'page_title'
            innerData.put("url", targetUrl);
            innerData.put("uploader_name", uploader);
            innerData.put("server_location_code", location); // Matches Backend 'server_location_code'
            innerData.put("upload_date", timestamp);
            innerData.put("html_hash", htmlHash); // Critical for Verification
            innerData.put("video_duration_minutes", duration);

            JSONObject payload = new JSONObject();
            payload.put("scraped_data", innerData); // WRAPPER REQUIRED BY SERIALIZER

            sendReport(payload);

            Thread.sleep(2000);

        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            driver.quit();
        }
    }

    // Helper function for Hashing
    public static String calculateSHA256(String base) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(base.getBytes(StandardCharsets.UTF_8));
            StringBuilder hexString = new StringBuilder();
            for (byte b : hash) {
                String hex = Integer.toHexString(0xff & b);
                if (hex.length() == 1) hexString.append('0');
                hexString.append(hex);
            }
            return hexString.toString();
        } catch (Exception ex) {
            return "HASH_ERROR";
        }
    }

    public static void sendReport(JSONObject json) {
        try {
            URL url = new URL("http://localhost:8000/api/report/");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setDoOutput(true);

            try (OutputStream os = conn.getOutputStream()) {
                byte[] input = json.toString().getBytes("utf-8");
                os.write(input, 0, input.length);
            }

            int responseCode = conn.getResponseCode();
            System.out.println("[+] Evidence Package Uploaded (HTTP " + responseCode + ")");
            
            if (responseCode != 201) {
                System.out.println("[-] Warning: Backend rejected the report. Check Server Logs.");
            }

        } catch (Exception e) {
            System.out.println("[-] CORE OFFLINE: " + e.getMessage());
        }
    }
}

File: SentinAL_Crawler/sentinel.py
import time
import json
import hashlib
import requests
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from bs4 import BeautifulSoup

# --- CONFIGURATION ---
TARGET_URL = "http://localhost:8080/pirate-site.html"
API_ENDPOINT = "http://127.0.0.1:8000/api/report/"

def calculate_hash(content):
    """Generates SHA-256 fingerprint of the HTML evidence."""
    return hashlib.sha256(content.encode('utf-8')).hexdigest()

def run_crawler():
    print(f"[*] Sentinel Crawler Activated. Target: {TARGET_URL}")

    # 1. Setup Headless Chrome
    chrome_options = Options()
    chrome_options.add_argument("--headless") 
    chrome_options.add_argument("--log-level=3") # Silence logs
    
    driver = webdriver.Chrome(options=chrome_options)

    try:
        # 2. Ingest
        driver.get(TARGET_URL)
        time.sleep(1) # Wait for render
        
        page_source = driver.page_source
        soup = BeautifulSoup(page_source, 'html.parser')

        # 3. Extract Metadata (The "Intelligence")
        title = soup.title.string if soup.title else "Unknown Asset"
        
        # Extract Meta Tags
        meta_uploader = soup.find("meta", attrs={"name": "uploader"})
        uploader = meta_uploader["content"] if meta_uploader else "Unknown"

        meta_location = soup.find("meta", attrs={"name": "server_location"})
        location = meta_location["content"] if meta_location else "XX"
        
        meta_duration = soup.find("meta", attrs={"name": "duration"})
        duration = int(meta_duration["content"]) if meta_duration else 0

        # 4. Generate Evidence Hash
        html_hash = calculate_hash(page_source)

        print(f"[+] TARGET ACQUIRED: {title}")
        print(f"[+] LOCATION: {location}")
        print(f"[+] DURATION: {duration} mins")
        print(f"[+] EVIDENCE HASH: {html_hash[:10]}...")

        # 5. Build Payload (Matching Backend Expectations)
        payload = {
            "scraped_data": {
                "page_title": title,
                "url": TARGET_URL,
                "uploader_name": uploader,
                "server_location_code": location,
                "video_duration_minutes": duration, # Critical for Integrity Check
                "html_hash": html_hash
            }
        }

        # 6. Send Report
        response = requests.post(API_ENDPOINT, json=payload)
        
        if response.status_code == 201:
            print(f"[+] SUCCESS: Evidence Package Uploaded (HTTP 201)")
        else:
            print(f"[-] FAILED: Backend rejected report. Status: {response.status_code}")
            print(response.text)

    except Exception as e:
        print(f"[-] CRITICAL FAILURE: {e}")
        # Hint: Make sure the mock site is running on port 8080!

    finally:
        driver.quit()

if __name__ == "__main__":
    run_crawler()